Fixes

samuelallan72 opened the pull request:
## Description

This fixes a subset of the reported issues found on the `wgu-release` branch (checked items have been addressed in this PR, unchecked items are still to discuss and work on):

- [x] 1. Progress slider not moving even though the audio is playing.
- [x] 2. "Download audio" link is pushed to the right and split over multiple lines
- [x] 3. Mute UI doesn't update when toggling mute
- [x] 4. Progress slider doesn't reach the end if an end time is set that's longer than the actual audio play time.
- [x] 5. Re-saving the xblock results in data loss.
- [x] 6. The multiple audio urls feature is confusing.
- [ ] 7. Sometimes the audio controls disappear, leaving a black box (I can't reliably reproduce this).
- [x] 8. transcript files are uploaded to /openedx/media/transcripts/ (they should be uploaded to a standard persistent storage)
- [x] 9. Uploaded transcript file is lost on re-save.
- [x] 10. The UI for changing the transcript needs improvement
- [x] 11. uploaded transcripts will override the previous one (ie. a transcript uploaded for another instance of the audio xblock) if they have the same file name
- [x] 12. If a new transcript is uploaded, the old one still remains on disk. (as an aside: do we want to support deleting a transcript as well?)
- [x] 13. the audio download link (if enabled) doesn't necessarily download - it may just open in the browser as an embedded audio player
- [x] 14. Podcast field is disabled if the audio file field is populated, and vice-verse. (UX needs improving)

## Supporting information

Private-ref: https://tasks.opencraft.com/browse/BB-9935

## Testing instructions

### Fix for: "Download audio" link is pushed to the right and split over multiple lines

- create an audio component in studio, with the "allow audio download" set to "yes"
- view the component in studio and verify the "Download audio" link is rendered on a single line, left aligned.
- view the component in the live course (student view), and verify the  "Download audio" link is rendered on a single line, left aligned as well.

### Fix for: issues with mute button and progress slider

- create an audio component in studio, with no start or end time set
- play the audio
- click the mute button a few times while it's playing
- verify the mute/unmute actions work as expected, and the UI correctly displays the mute/unmute status
- verify the play progress slider correctly displays the play progress as expected
- create another audio component in studio, with valid start and end times set (0 < start time < end time  < total duration)
- play the audio
- verify the audio begins playing at the set start time
- when it reaches the end time, verify the audio returns to the start time and stops playing
- verify the play progress bar displayed the progress as if the clipped area was the real start and end (ie. the progress bar starts at the far left, and ends at the far right)
- experiment with seeking the audio to before the start time and after the end time
- there are other minor known issues with this, but for this purpose, verify the audio is bounced to the set start time whenever you go past the set bounds.

### Fix for: podcast/audio file field disabled if the other is set

- create an audio component in studio
- set it to "audio file" type and upload an audio file and save it
- edit the audio component again
- switch to the podcasts type
- verify you can enter a podcast embed url and successfully save it.

Note that switching between audio file and podcasts will clear the data from the now-inactive tab. Not sure if that is desired or not. It makes for a smooth experience, but could result in unexpected data loss if a user accidentally switches tabs before saving.

### Fix for: re-saving the xblock results in data loss

- create an audio component in studio, setting several fields
- save
- click to edit the block again
- don't modify any fields
- click "save"
- verify that nothing has changed
- try editing and saving a few more times, with different data inputs
- verify data is saved as expected.

### Fix for: confusing multiple urls for audio file

- create an audio component in studio
- in the editor, consider the placeholder and help text for the audio file urls:
- verify the purpose is clear

### Fix for: transcript uploaded to ephemeral path

- create an audio component in studio
- in the audio component editor, upload a transcript file
- verify the file is uploaded to the default storage location (in a Tutor devstack, this should be /openedx/media/ in the cms and lms shell (volume backed) - this matches the MEDIA_ROOT django setting)
- verify the transcript displays when playing the audio component

### Fix for: transcript lost on re-save

- use the audio component with an uploaded transcript file (from the previous step)
- edit the component in studio, but don't make any changes
- save the component
- verify the transcript is still present

### Fix for: bad UI for transcript selector

- use the audio component with an uploaded transcript file (from the previous step)
- use the transcript selector (the 'CC' button on the media player)
- verify the controls are all visible (not cut off at the top as it was before), readable, and accessible

See screenshot of it fixed for reference:

<img width="382" height="285" alt="1754976786" src="https://github.com/user-attachments/assets/34bafdf8-d826-401d-ba29-4901cf0f3607" />

### Fix for: if a new transcript is uploaded, the one one is still kept on disk

- use the audio component with an uploaded transcript file (from the previous step)
- locate the transcript file on disk (somewhere in /openedx/media/ in a Tutor devstack)
- edit the audio component, uploading a new transcript file
- verify the old transcript file is no longer present on disk
- verify the new transcript file saved successfully, and is visible when the audio is played

### Delete transcript functionality

- use the audio component with an uploaded transcript file (from the previous step)
- edit the audio component: delete the transcript file (using the Delete checkbox)
- save
- verify the transcript file is no longer present on disk
- verify no transcript is available when playing the audio

### Fix for: audio download link doesn't download

NOTE: there isn't a reliable fix for this - it's not possible here to force the browser to download the file instead of embedding it. We make some improvements here though:
- Add the "download" attribute, to ask the browser to download instead.
      Note that this only works for same origin urls.
- Add the target="_blank" attribute to open in a new tab.
      This avoids getting navigated away and losing your work.

Testing instructions:
- create an audio component with "allow audio download" set to "yes"
- save and view the audio
- click the "download audio" link
- verify the file opens in a new tab/window (it will likely be embedded by the browser still)
- verify the link has the "download" attribute set (check the html source).
- (optional): The download attribute only works for same-origin files - you can test the download attribute by uploading an audio file to studio, and adding that url to the audio component. Click the "download audio" link and verify the browser downloads the file instead of displaying it embedded.

### Fix for: progress slider doesn't reach the end if end time is longer than the audio duration

- create an audio component
- set an end time to something very large, much larger than the audio duration (eg. 500 hours)
- play the audio
- verify the progress slider updates as expected (ie. proportional to the actual play time)
- also test with no end time set, and with an end time somewhat shorter than the actual audio duration

## Deadline

None

## Other information

The `wgu-release` branch (that this PR targets) reflects the actual state of what is deployed. This `wgu-release` branch includes cherry picks of the following PRs (at time of writing):

- #1
- #3

samuelallan72 committed:
fix: improve rendering of "Download audio" link

Before, it was pushed to the right
and split over multiple lines in the studio view.

Now it displays on a single line, aligned to the left,
and is consistent between student and studio views.

Private-ref: https://tasks.opencraft.com/browse/BB-9935

samuelallan72 committed:
fix: Fix controls when end time is unset

Previously, without an end time set,
there were issues with the mute button and the progress slider.

Fix this by improving how the endTime is calculated in the frontend,
and checking if endTime is set before running any of the custom logic
for clipping the play based on start and end times.

Private-ref: https://tasks.opencraft.com/browse/BB-9935

samuelallan72 committed:
fix: Improve UX and data consistency on form

This fixes a range of issues:
- There were two form elements on the page,
  which was causing double saving and lost data.
- Submitting the form caused a page refresh on success,
  which was unnecessary.
- Switching between podcast and audio file mode was confusing,
  because fields were disabled.
- Editing previously saved start and end times was broken,
  because of a missing conversion between seconds and MM:SS formats.

This improves the situation by:
- Reorganising the html layout so there's only a single form.
- Clearing the hidden fields when the audio type is switched.
- Removing the force page refresh on success.
- Adding conversion from seconds to MM:SS

samuelallan72 committed:
fix: Clarify purpose of multiple URLs for audio

samuelallan72 committed:
fix: improve handling transcripts

- Upload transcripts to a configurable Django storage.
- Display if a file is currently saved in the edit UI.
- Support deleting an uploaded transcript file.
- Fix a bug where an existing uploaded transcript file is deleted
  if no file is selected when saving the form.
- Improve display of the caption selector in the media player.

xitij2000 commented on audio/public/js/audio.js:
```diff
@@ -5,21 +5,16 @@ function AudioBlock(runtime, element) {
-    const endTime = parseFloat(audioElement.dataset.endTime) || audioElement.duration;
+    // these default to 0.0 if unset
+    const startTime = parseFloat(audioElement.dataset.startTime);
+    const endTime = parseFloat(audioElement.dataset.endTime);
```
They may default to zero, but it's possible that an older version of the block didn't and would cause an error here. I am not sure there is any advantage in making this less explicit.

samuelallan72 replied:
Ah I can update the comment perhaps - the backend logic defaults them to zero if unset, so this should be fine. I can double check though. :+1:

xitij2000 commented on audio/public/js/audio.js:
```diff
@@ -5,21 +5,16 @@ function AudioBlock(runtime, element) {
         success: function(mediaElement, originalNode) {
             mediaElement.setCurrentTime(startTime);
 
-            Object.defineProperty(mediaElement, 'duration', {
-                get: function() {
-                    return endTime - startTime;
-                }
-            });
```
Why remove this?

samuelallan72 replied:
The previous logic for managing endTime and duration was broken for a few reasons, so I rewrote most of it. :)

xitij2000 replied:
My point is the the duration is still used in places, should the code be fixed here so you can use the duration property?

samuelallan72 replied:
Ah right - where I'm using the duration property, it's the [duration property included in the HTMLMediaElement web api](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/duration). :)

samuelallan72 committed:
fix: improve download audio link

Browsers tend to open audio file links as embedded audio in the browser.
This helps improve the situation:
- Add the "download" attribute, to ask the browser to download instead.
  Note that this only works for same origin urls.
- Add the target="_blank" attribute to open in a new tab.
  This avoids getting navigated away and losing your work.

samuelallan72 committed:
fix: limit end time to within audio duration

It previously wasn't limited by audio duration,
which caused the player progress slider to not update as expected
if the end time was longer than the audio.

samuelallan72 commented on audio/public/css/audio.css:
```diff
@@ -1,6 +1,7 @@
 .audio-xblock {
   display: block;
   margin-bottom: 4rem;
+  margin-top: 4rem;
```
This is a rather hacky way to ensure the caption selector isn't cut off at the top... not sure if there's a better way.

xitij2000 commented on audio/audio.py:
```diff
@@ -120,15 +148,39 @@ def studio_submit(self, request, suffix=''):
+        storage = get_storage_backend()
+        will_upload_transcript_file = 'transcript_file' in data and hasattr(data.get('transcript_file'), 'file')
+
+        if data.get("delete_transcript_file", "") == "on" or will_upload_transcript_file:
+            if self.transcript_file and storage.exists(self.transcript_file):
+                storage.delete(self.transcript_file)
```
I think this should be a separate operation entirely. i.e a new API endpoint to delete the transcript that is called by a button the frontend for the same.

samuelallan72 replied:
Hmm perhaps - could you see what you think after you use it in the UI - I think it will make more sense then. I modelled it after the Django admin file field UX.

xitij2000 replied:
I've seen it, however, I don't like the idea of tying the delete operation to a save, since it means that you now have a two-step process to swap the subtitles. 

Perhaps it can auto-delete when you upload a new one and that's it.

samuelallan72 replied:
> have a two-step process to swap the subtitles.

I'm not sure what you mean. It's still a single step for the user to swap transcripts. The types of interactions could look like this:

- "I want to upload a transcript": click the browse files and select one. Then save the form. (the transcript will be uploaded and saved)
- "I want to replace the transcript with a new one": click the browse files and select one. Then save the form. (the new transcript will saved, and the old one cleaned up; note that in this case the effect is the same regardless of whether the user clicks the delete checkbox next to the existing transcript)
- "I no longer want a transcript here": click the "delete" checkbox. Then save the form. (the transcript will be deleted)

This feels cohesive and consistent, as everything is in the same form, and nothing is updated on the backend until the user clicks Save.

xitij2000 replied:
I disagree about it feeling cohesive and consistent. As you explain in the second point, the button has no effect if you upload one subtitle to replace another. Having it be a button is, I feel, a more consistent experience.

samuelallan72 replied:
> Since the file deletion feature isn't explicitly requested perhaps we can leave out that UI till it's discussed.

@xitij2000 are you ok if we merge the file deletion as-is, and file a follow-up ticket to investigate the UX? I'm not really comfortable removing the ability to delete an uploaded transcript, as this could cause issues in the case where a user accidentally uploads the wrong file and needs to delete it.

samuelallan72 replied:
In the interest of time, I merged as-is. I can address it in a follow up PR if you like, or I can open a follow-up ticket to investigate changes further. :)

xitij2000 replied:
There is a way to delete the transcript, which is to upload the correct transcript which will clear the old file. You can also delete it direct from S3 in case of an emergency.

samuelallan72 replied:
That is true :+1:

samuelallan72 replied:
Sorry if I jumped the gun with merging earlier. Please see follow up PR: https://github.com/open-craft/xblock-audio/pull/9 which removes the delete checkbox. It will take more work to discuss UI/UX options and implement a different method, so I think we can defer that to a future ticket.

ali-hugo replied:
@samuelallan72 @xitij2000 My thoughts on the UX for deleting an uploaded file:

- **Have an explicit “remove” option:** Users should be able to remove a file without needing to replace it with another or delete it from S3.
- **Use a button, not a checkbox:** A dedicated button keeps the action clear and avoids tying deletion to a “save” step.
- **Wording / iconography:** “Remove” is more commonly used than “delete.” This could be a labelled button, or a simple “X” icon next to the file name.
- **Examples for inspiration:**
  - [Good example of file upload with a clear remove option](https://www.designsystem.qld.gov.au/components/file-upload)
  - [Demo showing the “X” approach](https://carbondesignsystem.com/components/file-uploader/usage/)
  
Interested to hear what you think! Let me know if you have any questions / concerns.

samuelallan72 replied:
Thanks @ali-hugo , this makes a lot of sense, and lines up with what you were saying @xitij2000 . :+1: :+1: 

I'm not sure how much time we have left on this ticket, or if we can open a new ticket for the future to implement this. We can discuss on the ticket. :)

samuelallan72 replied:
Follow up PR with the new UI has been created at https://github.com/open-craft/xblock-audio/pull/10 for review. :)

xitij2000 commented on audio/audio.py:
```diff
@@ -120,15 +148,39 @@ def studio_submit(self, request, suffix=''):
+            transcript_file = data['transcript_file']
+
+            # generate a safe path for the transcript file
+            loc = self.location
```
This is the usage_key right? So call it that to avoid confusion sincee location can also be be a path.

samuelallan72 replied:
I'm not sure - I copied this from [edx-sga](https://github.com/mitodl/edx-sga/blob/720b3106a275896929d1433b52eb9ebc57892c8c/edx_sga/sga.py#L970-L970) - `self.location` is not defined in this codebase, so it must be a standard xblock attribute.

xitij2000 replied:
It is, however my concern is about the name `loc` not the property. I think the new alternative is `self.scope_ids.usage_id` I think, but instead of loc, you could call it `blk_id` or `usage_id`

samuelallan72 replied:
Ok nice, thanks. I'll take a look at that. :)

samuelallan72 replied:
I found it, and updated to use usage_key (which is scope_ids.usage_id as you thought). :) I also passed it through django's `get_valid_filename` to avoid issues with the `+` and `:` characters in the usage keys - not sure if necessary though.

xitij2000 commented on audio/audio.py:
```diff
@@ -120,15 +148,39 @@ def studio_submit(self, request, suffix=''):
+
+            # generate a safe path for the transcript file
+            loc = self.location
+            file_path = f"{loc.org}/{loc.course}/{loc.block_type}/{loc.block_id}/{uuid4()}.vtt"
```
You should use the extension of the original file. Additionally, I see no point in using UUID here. The cobination of org, course, block type, and block id is unique to this block instance, and keeping the original file name might be better.

samuelallan72 replied:
I used UUID and hardcoded extension to keep everything consistent. I can definitely use the original file name and extension here though if you think that's better - although I would normalise the name to avoid security issues.

xitij2000 replied:
Rather than UUID, you can use either DJango's `get_random_string` function to generate a small file suffix to ensure things are unique, or just add the epoch time which will always increase.

samuelallan72 replied:
I updated it to keep the original filename, but passing it through django's `get_valid_filename` to avoid issues.

xitij2000 commented on audio/audio.py:
```diff
@@ -120,15 +148,39 @@ def studio_submit(self, request, suffix=''):
+            try:
+                return Response(
+                    app_iter=iter(partial(storage.open(path).read, BLOCK_SIZE), b""),
+                    content_type='text/vtt',
```
Likewise, content type for subtitles could be srt as well right?

samuelallan72 replied:
Only WebVTT is supported in browsers, so we can safely only accept vtt.

xitij2000 replied:
Okay, makes sense then!

xitij2000 reviewed (commented):
I've done a code pass, will test on an instance next.

samuelallan72 commented:
@xitij2000 thanks for your review comments! I've responded. :slightly_smiling_face: I'll hold off on making any code changes until you've tested functionality on an instance.

xitij2000 reviewed (approved):
:+1: Looks good, just a few issues that I pointed out. 

Since the file deletion feature isn't explicitly requested perhaps we can leave out that UI till it's discussed.

- [x] I tested this: tested fixes on devstack
- [x] I read through the code
- [na] I checked for accessibility issues
- [x] Includes documentation

samuelallan72 committed:
fix: improve file path for uploaded transcript

- base the filename on the actual file's filename (safely)
- use usage_key instead of location
